python/cuspatial/cuspatial/geometry/geocolumn.py

```diff
@@ -0,0 +1,471 @@
+
+    @property
+    def points(self):
+        return self._geo._points
+
+    @property
+    def multipoints(self):
+        return self._geo._multipoints
+
+    @property
+    def lines(self):
+        return self._geo._lines
+
+    @property
+    def polygons(self):
+        return self._geo._polygons
+
+    def _dump(self):
+        return (
+            f"POINTS\n"
+            f"{self.points._repr__()}\n"
+            f"MULTIPOINTS\n"
+            f"{self.multipoints._repr__()}\n"
+            f"LINES\n"
+            f"{self.lines._repr__()}\n"
+            f"POLYGONS\n"
+            f"{self.polygons._repr__()}\n"
+        )
+
+    def __repr__(self):
+        return (
+            f"GeoColumn\n"
+            f"{len(self.points)} POINTS\n"
+            f"{len(self.multipoints)} MULTIPOINTS\n"
+            f"{len(self.lines)} LINES\n"
+            f"{len(self.polygons)} POLYGONS\n"
+        )
+
+    def copy(self, deep=True):
+        """
+        Create a copy of all of the GPU-backed data structures in this
+        GeoColumn.
+        """
+        result = GeoColumn(self._geo.copy(), self._meta.copy(), self.data)
+        return result
+
+    def _copy_type_metadata(self: T, other: ColumnBase) -> ColumnBase:
+        self._data = other.data
+        return self
+
+
+class GeoColumnLocIndexer:
+    """
+    Not yet supported.
+    """
+
+    def __init__(self):
+        # Todo: Easy to implement with a join.
+        raise NotImplementedError
+
+
+class GeoColumnILocIndexer:
+    """
+    Each row of a GeoSeries is one of the six types: Point, MultiPoint,
+    LineString, MultiLineString, Polygon, or MultiPolygon.
+    """
+
+    def __init__(self, sr):
+        self._sr = sr
+
+    def __getitem__(self, index):
+        if not isinstance(index, slice):
+            mapped_index = int(self._sr.values[index])
+            return self._getitem_int(mapped_index)
+        else:
+            raise NotImplementedError
+            # This slice functionality is not ready yet
+            # return self._getitem_slice(index)
+
+    def _getitem_int(self, index):
+        type_map = {
+            "p": GpuPoint,
+            "mp": GpuMultiPoint,
+            "l": GpuLineString,
+            "ml": GpuMultiLineString,
+            "poly": GpuPolygon,
+            "mpoly": GpuMultiPolygon,
+        }
+        return type_map[self._sr._meta.input_types[index]](self._sr, index)
+
+
+class GpuGeometry:
+    def __init__(self, source, index):
+        """
+        The base class of individual GPU geometries. This and its inheriting
+        classes do not manage any GPU data directly - each GpuGeometry simply
+        stores a reference to the GeoSeries it is stored within and the index
+        of the geometry within the GeoSeries. Child GpuGeometry classes
+        contain the logic necessary to serialize and convert GPU data back to
+        Shapely.
+        """
+        self._source = source
+        self._index = index
```

Comment:
The design of this class makes me really uneasy. In looking at this class and the subclasses, I assumed these were similar to `cudf.Scalar` where this class would be used to represent an element of data. If that is the case, you could imagine someone using these in operations like scatters and comparisons, where holding a reference to its source `GeoSeries` can be hugely memory inefficient, and the holding of the index where it is within the `GeoSeries` seems like a separation of concerns.

Based on my understanding of the code below, we're doing this to be able to convert to shapely objects per element. If we need to do this, then I think we should do a bulk copy to the host instead of having methods to convert a GPU element to a host shapely object. Then we could more comfortable iterate over the bulk host object and convert things to shapely objects in that iteration.

Reply:
I think this does what you're asking for, in the latter case, though the class name could be improved to reflect that. At this time there is a very limited implementation of the former case in the `.data` buffer of the `NumericalColumn` that `GeoColumn` is based on. The `.data` buffer is used as an integer mapping, originally the identity, from rows in the `GeoColumn` to geometries stored in the `GeoArrowBuffers`. It can be reordered, such as via `DataFrame.sort_value`, which produces a different ordering of the rows without modifying the underlying data.

The `GpuGeometry` classes contain the necessary logic to extract all of the relevant coordinates and geometries for Points, MultiPoints, LineStrings, MultiLineStrings, Polygons, and MultiPolygons. The most significant application of this logic is for creating `Shapely` classes that can be inserted into a `GeoPandas` dataframe.

`cuspatial` doesn't offer any single-geometry-wise operations at this time - our GPU accelerated algorithms are all batch jobs on many geometries. Eventually it'll be nice to be able to completely rearrange and reconstruct `GeoArrowBuffers` based on the indices of individual geometries, but it is not necessary at this time.

The `GpuGeometry` classes are only used within `cuspatial` code after `to_host` is called, moving all of the `GeoArrowBuffers` arrays to host. They were originally `camelCased` class names to reflect they are "private" classes, but I'm open to a better scheme for naming and hiding them from accidental misuse. Someone attempting to write their own `cuspatial` library code might use them incorrectly. One solution would be to examine the host/device status of the buffer in question when `to_shapely` is called and raising an error if it hasn't been copied to host first.

Reply:
I'm going to rename the classes to be something along the line of `ShapelySerializer`, which should resolve the issue conceptually. :)